src/main.rs

```diff
@@ -71,203 +84,227 @@ impl Player {
-            description.push_str("\nYou hear a rustling.");
+    /// Index of neighboring room given by user
+    /// `destination`, else an error message.
+    fn parse_room(
+        &self,
+        destination: &str,
+        current_room: usize,
+    ) -> Result<usize, String> {
+        let dest: usize = destination.trim().parse().map_err(
+            |e: std::num::ParseIntError| e.description().to_string(),
+        )?;
+
+        // Check that the given destination is the id of a linked room.
+        if self.rooms[current_room]
+            .neighbors
+            .iter()
+            .all(|&r| r != dest)
+        {
+            return Err("room is not a neighbor".to_string());
         }
-        if self.is_danger_nearby(room, Danger::Wumpus) {
-            description.push_str("\nYou smell something terrible nearby.");
-        }
-
-        description.push_str(&format!("\nExits go to: {}",
-                                      self.rooms[room].neighbours
-                                          .iter()
-                                          .map(|n| n.get().unwrap().to_string())
-                                          .collect::<Vec<String>>()
-                                          .join(", ")));
 
-        description
+        Ok(dest)
     }
 
-    fn is_danger_nearby(&self, room: usize, danger: Danger) -> bool {
-        self.rooms[room].neighbours.iter().find(|n| {
-            self.rooms[n.get().unwrap()]
-                .dangers.contains(&danger)
-        }).is_some()
-    }
-
-    fn parse_room(&self, destination: &str, current_room: usize) -> Result<usize, ()> {
-        let destination: Result<usize, _> = destination.parse();
-
-        // check that the given destination is both a number an the number of a linked room
-        if let Ok(room) = destination {
-            if self.rooms[current_room].neighbour_ids().contains(&room) {
-                return Ok(room);
+    fn empty_room(&mut self) -> usize {
+        // Number of rooms that should be empty. This should
+        // maybe be replaced with a counting loop.
+        let nempty = self.rooms.len() - BATS - PITS - 1;
+        let mut n = self.rng.gen_range(0, nempty);
+        for (i, r) in self.rooms.iter().enumerate() {
+            if !r.wumpus && r.contents == Empty {
+                if n == 0 {
+                    return i;
+                }
+                n -= 1;
             }
         }
-
-        Err(())
+        panic!("internal error: could not find empty room");
     }
 }
 
-///////////////
-// MAIN LOOP //
-///////////////
-
+/// Current game state.
 enum Status {
     Normal,
     Quitting,
     Moving,
     Shooting,
 }
+use Status::*;
 
 fn main() {
-    let rng = Rc::new(RefCell::new(rand::thread_rng()));
-    let mut maze = Maze::new(rng.clone());
-    let mut player = Player::new(maze.rnd_empty_room());
-    let mut status = Status::Normal;
+    let (mut maze, player_loc) = Maze::new();
```

Comment:
I'm aware that having the *rng* shared among the main function and the maze makes the code more complex, but generally it's a good design the use a single "randomness generator" instance used throughout the whole code, since it can be seeded in a single point and have the whole code work in a "pseudo-random" way, which is useful for testing and debugging.

Reply:
Are you suggesting that something change here? There is a single prng that is shared throughout the whole code. I chose to have the Maze own the prng, since it's notionally part of the game state and since it makes it convenient to pass the prng around… The prng could be owned by `main()` and passed explicitly by mut reference everywhere if you feel that's cleaner.

I did remove a clone of the prng in the original code. While it is legit to clone the prng, it's usually better not to if it can be avoided: it's expensive (doesn't matter here) and has the potential to mess with independence in bad prng implementations. (You can't just clone the prng state, else now you have two generators that will both produce the same numbers. So you have to reinit the state of the "new" generator somehow.)